lib/srv/regular/proxy.go

```diff
@@ -260,12 +263,46 @@ func (t *proxySubsys) proxyToHost(ctx context.Context, ch ssh.Channel, clientSrc
 	}
 
 	go func() {
-		t.close(utils.ProxyConn(ctx, ch, conn))
+		t.close(utils.ProxyConn(ctx, &checkedPrefixReader{
+			ReadWriteCloser: ch,
+			// SSH connection MUST start with "SSH-2.0" bytes according to https://datatracker.ietf.org/doc/html/rfc4253#section-4.2
+			requiredPrefix: []byte("SSH-2.0"),
+		}, conn))
 	}()
 
 	return nil
 }
 
+// checkedPrefixReader checks that read data has the specified prefix.
+type checkedPrefixReader struct {
+	io.ReadWriteCloser
+	requiredPrefix []byte
+
+	// upstreamReader reads from the underlying reader.
+	upstreamReader io.Reader
+}
+
+func (c *checkedPrefixReader) Read(b []byte) (int, error) {
+	// connection was already checked, forward upstream:
+	if c.upstreamReader != nil {
+		return c.upstreamReader.Read(b)
+	}
+
+	// make sure first bytes are equal to the required prefix
+	reader := bufio.NewReader(c.ReadWriteCloser)
+	data, err := reader.Peek(len(c.requiredPrefix))
```

Comment:
_technically_ this prefix could be split across multiple packets.  Given that this is the start of the data it's highly unlikely unless someone was doing it purposefully.  But it is a case to potentially consider or comment that it's just required that the prefix is contained within a single packet.

Reply:
`(*bufio.Reader).Peek` will read from the inner reader until it's got enough data to fulfill the peek or until it gets a read error, so we're not relying on any underlying framing here.